{{cookiecutter.project_slug}}/server/{{cookiecutter.project_slug}}/settings.py

```diff
@@ -215,25 +217,33 @@
 #
 # Static files (CSS, JavaScript, Images)
 #
-{% if cookiecutter.client_app != "React" -%}
-STATIC_ROOT = os.path.join(BASE_DIR, "static")
-MEDIA_ROOT = os.path.join(BASE_DIR, "media-files")
-{% else %}
-STATIC_ROOT = os.path.join(BASE_DIR, "..", "client", "build")
-MEDIA_ROOT = os.path.join(BASE_DIR, "..", "client", "build", "static")
-{% endif -%}
+# Django will create directories for STATIC_ROOT and MEDIA_ROOT.
+# Static files are things like JS, CSS, and images. Media files are
+# user-uploaded files. By default, media files are stored on the local
+# file system when uploaded. This is fine for development, but but when
+# on Heroku, you must use an external system like AWS S3, because the
+# Heroku file system is destroyed during each deployment.
+#
+
+# Static files will be collected into 'static' when `manage.py collectstatic` is run
+STATIC_ROOT = os.path.join(BASE_DIR, "..", "static")
+MEDIA_ROOT = os.path.join(BASE_DIR, "..", "media-files")
 
+# Static and media files will be served from under these paths.
 STATIC_URL = "/static/"
 MEDIA_URL = "/media/"
 
 {%- if cookiecutter.client_app != 'None' %}
 {%- if cookiecutter.client_app == 'Vue3' %}
+# Django will look for client-side build files in this directory
 STATICFILES_DIRS = [
     os.path.join(BASE_DIR, "../client/dist/static"),
 ]
 {% elif cookiecutter.client_app == 'React' %}
+# Django will look for client-side build files in this directory
 STATICFILES_DIRS = [
-    os.path.join(STATIC_ROOT, "static"),
+    os.path.join(BASE_DIR, "..", "client", "build"),
+    os.path.join(BASE_DIR, "..", "client", "build", "static"),
```

Comment:
^ @OKitsMikael for your review: These are the settings that worked on our recent client project for React. Note that we're importing both the outer `client/build` folder and the inner `client/build/static`. This means files in `client/build/static` are copied to the `STATIC_ROOT` twice. On the first pass they are copied to `static/static` and the second time they are copied to `static/`.

This is a hack. I had to do this, because Django Whitenoise was otherwise not able to resolve the path to some font files referenced in the CSS. Those font files were in `/static/static/media/`, but it was looking in `/static/media/`, so copying the `client/static/` folder to the top level of `STATIC_ROOT` resolved the issue.

Confusing, I know. Let know if that makes sense.

Reply:
@whusterj the bit about copying files to `STATIC_ROOT` twice seems a little odd to me, my understanding was that Django would look in this list of directories for static files once `collectstatic` runs and dumps the static files into `STATIC_ROOT`, thus merging any assets on the server side into the same folder that's used to hold the client-side static files.

I also see that on L228, you've changed `STATIC_ROOT`, so the files will no longer be put into `client/build/static` but instead `/static`. Not sure what else this change will affect.

I think we could maybe play with the `STATIC_URL` and `MEDIA_URL` so that we don't copy the same files into multiple directories if that's indeed what's going on. I'll generate a test project, and load it into a review app to make sure that fonts, images, and favicons all work properly.

Reply:
@OKitsMikael If you change this file: https://github.com/thinknimble/tn-spa-bootstrapper/blob/main/buildpack-run.sh
You should be able to have it generate a react app for the review app on this PR to ease testing of these changes. (I've been thinking about how to automate this going forward)

Reply:
Your description of `collectstatic` sounds correct, and the merging of client and server-side files is effectively what happens with our config.

Maybe a more thorough breakdown will clarify things: `STATICFILES_FINDERS` lists the classes used to find static files. First up is `FileSystemFinder`, which looks in the `STATICFILES_DIRS` and copies those over to `STATIC_ROOT`, followed by `AppDirectoriesFinder`, which grabs the contents of the `static/` directory in each Django app.

1. So first, it copies the contents of `client/build` to `STATIC_ROOT`
2. Second, it copies the contents of `client/build/static` to `STATIC_ROOT`
3. Third, it copies the contents of the `static` directory from each Django app to `STATIC_ROOT`

So the contents of `client/build/static` ends up existing in two places: `STATIC_ROOT/static` after step one and `STATIC_ROOT/` after step 2. Again, I only did this because of an error thrown by Django Whitenoise when running `collectstatic`. It said it could not resolve a font name that was referenced in a CSS file and existed in `client/build/static/media/`. It did not like the file being in `$STATIC_ROOT/static/media/`, but it was fine with `$STATIC_ROOT/media/`, and the second step above accomplishes this.

I think this may actually be a Whitenoise bug, but haven't had time to really look at it. Ideally, we should only need step 1 from the above list.

>I also see that on L228, you've changed STATIC_ROOT, so the files will no longer be put into client/build/static but instead /static. Not sure what else this change will affect.

It shouldn't affect anything else. The choice of directory name and location is mostly arbitrary. The only two limitations I can think of are that you don't want to overwrite existing directories, and that the process that runs Django must have permission to access it. So AFAIK, this won't have any side effects.

Also, to be clear, it's not set to the root directory `/static`. It's `$BASE_DIR/../static/`, where `$BASE_DIR` is the `server/` directory. So this puts the `static` directory at the top level of the _project_ (that is, "above" `client` and `server`). On Heroku, for instance, this ends up resolving to `/app/static`.

>I think we could maybe play with the STATIC_URL and MEDIA_URL so that we don't copy the same files into multiple directories if that's indeed what's going on

I don't think changing these settings would make a difference. These are just the URLs or path names Django uses to generate links to static and media files for use on the front end.

One other note, `MEDIA_ROOT` should never have been set to `client/build/static`, since it [is intended for user-uploaded files](https://docs.djangoproject.com/en/4.1/ref/settings/#std-setting-MEDIA_ROOT), not static files. So it should be kept entirely separate, and in prod probably not kept on the filesystem at all.

Reply:
@whusterj Thanks for the thorough explanation! This gives me a much clearer understanding of the process. There shouldn't be many files in `client/build/static`, so I can see how copying them twice isn't ideal, but it's not that much of an issue.

Thanks for fixing `MEDIA_ROOT`. We definitely hit the issue with user-uploaded media files getting deleted on deploy (which we resolved by using a Heroku addon for user-uploaded images).